File: DS-Algo/inflearn_algo_basic/11-12/12.cpp
#include <iostream>
#include <vector>
using namespace std;
// #define DEBUG

/*
    @brief : count total using digit from 1~n (e.g. n =3 1,2,3 return 3)
    @param : n(int) -> input num
    @return : cnt(int)
*/
int countAllDigit(int n)
{
    int temp_n = n;
    int ten = 1, mul = 1;
    int cnt=0;

    while (1)
    {
        if (temp_n / 10 != 0)
        {
            cnt += ((ten * 10) - ten) * mul;
            ten = ten * 10;
            mul++;
            temp_n /= 10;
        }
        else
        {
            cnt += ((n + 1) - ten) * mul;
            break;
        }
    }

    return cnt;
}

int main()
{
    int N;
    cin >> N;
    if (N < 3 && N > 100000)
    {
        printf("[ERROR] invalid input (3<=N<=100000000)");
        exit(-1);
    }

#ifdef DEBUG
    printf("[INFO] N = %d\n", N);
    printf("[INFO] countAllDigit of %d : %d\n", N, countAllDigit(N));
#endif

    cout << countAllDigit(N);

    return 0;
}

File: DS-Algo/inflearn_algo_basic/11-12/12.py
import logging

logging.basicConfig(level=logging.WARNING, format='[%(levelname)s] %(message)s')

def count_all_digit(n : int):
    '''
        cout all digit 1~n (e.g. input : 3 -> 1,2,3 return 3)

        Args:
            n(int) 
        Returns:
            cnt(int)
    '''
    cnt = 0
    mul = 1
    digit_cnt = 9
    digit_sum = 0

    while(digit_sum + digit_cnt < n):
        cnt += digit_cnt * mul
        digit_sum += digit_cnt
        mul += 1
        digit_cnt*=10
    
    cnt += (n-digit_sum) * mul
    return cnt

N = int(input())

logging.info(f'input N : {N}')

print(count_all_digit(N))